### PR TITLE
Headers implementation

### PR DESCRIPTION
**Pull request by antmat**
header support for native framework

### CODE REVIEW

File: include/cocaine/framework/detail/decoder.hpp [hunk 1]
```diff
@@ -31,7 +33,8 @@ namespace detail {
 /// \internal
 struct decoder_t {
     typedef decoded_message message_type;
-
+    io::header_table_t header_table;
+    msgpack::zone zone;
```
Comment (3Hren) [inline include/cocaine/framework/detail/decoder.hpp]: Blank line.

File: include/cocaine/framework/message.hpp [hunk 2]
```diff
@@ -21,6 +21,8 @@
 #include <stddef.h>
 
 #include <boost/none_t.hpp>
+#include <boost/optional.hpp>
+#include <cocaine/rpc/asio/header.hpp>
```
Comment (3Hren) [inline include/cocaine/framework/message.hpp]: Blank.

File: include/cocaine/framework/message.hpp [hunk 3]
```diff
@@ -35,13 +37,7 @@ class decoded_message {
...
     explicit decoded_message(boost::none_t);
 
-    /// Constructs a message object from its raw parts.
-    ///
-    /// \pre the message should be valid MessagePack'ed Cocaine message, otherwise the behavior is
-    /// undefined.
-    /// \throws std::bad_alloc if unable to allocate the required memory chunk.
-    decoded_message(const char* data, std::size_t size);
-
+    decoded_message(msgpack::object, std::vector<char> storage, std::vector<io::header_t> headers);
```
Comment (3Hren) [inline include/cocaine/framework/message.hpp]: Don't do that. Restore and actualize the documentation.

File: include/cocaine/framework/sender.hpp [hunk 4]
```diff
@@ -53,11 +53,19 @@ class basic_sender_t {
     template<class Event, class... Args>
     auto
     send(Args&&... args) -> task<void>::future_type {
-        return send(io::encoded<Event>(id, std::forward<Args>(args)...));
+        namespace ph = std::placeholders;
```
Comment (3Hren) [inline include/cocaine/framework/sender.hpp]: Use full `std::placeholders` namespace in header files.

File: include/cocaine/framework/session.hpp [hunk 5]
```diff
@@ -74,20 +74,19 @@ class session {
     typename task<channel<Event>>::future_type
     invoke(Args&&... args) {
         namespace ph = std::placeholders;
```
Comment (3Hren) [inline include/cocaine/framework/session.hpp]: Oops, drop this alias at once :)

File: include/cocaine/framework/sender.hpp [hunk 6]
```diff
@@ -53,11 +53,19 @@ class basic_sender_t {
...
 private:
-    auto send(io::encoder_t::message_type&& message) -> task<void>::future_type;
+    template<class Event, class... Args>
+    static
+    io::encoder_t::message_type
+    encode(std::uint64_t span, io::encoder_t& encoder, Args&... args) {
+        return encoder.encode<Event>(span, std::forward<Args>(args)...);
+    }
+
+    auto send(const std::function<io::encoder_t::message_type(io::encoder_t&)>& message_getter) -> task<void>::future_type;
```
Comment (3Hren) [inline include/cocaine/framework/sender.hpp]: Consider declaring typedef for function.

File: src/basic_session.cpp [hunk 7]
```diff
@@ -178,7 +178,7 @@ basic_session_t::cancel() {
 }
 
 framework::future<basic_session_t::invoke_result>
-basic_session_t::invoke(std::function<io::encoder_t::message_type(std::uint64_t)> encoder) {
+basic_session_t::invoke(const std::function<io::encoder_t::message_type(std::uint64_t, io::encoder_t&)>& message_getter) {
```
Comment (3Hren) [inline src/basic_session.cpp]: Why have you renamed the variable?

File: src/basic_session.cpp [hunk 8]
```diff
@@ -192,7 +192,7 @@ basic_session_t::invoke(std::function<io::encoder_t::message_type(std::uint64_t)
     auto rx    = std::make_shared<basic_receiver_t<basic_session_t>>(span, shared_from_this(), state);
 
     channels->insert(std::make_pair(span, std::move(state)));
-    return push(encoder(span))
+    return push(std::bind(message_getter, span, std::placeholders::_1))
```
Comment (3Hren) [inline src/basic_session.cpp]: Here you can use `ph::*`.

File: src/decoder.cpp [hunk 9]
```diff
@@ -32,23 +36,26 @@ size_t decoder_t::decode(const char* data, size_t size, message_type& message, s
...
 
     msgpack::object object;
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
-    std::auto_ptr<msgpack::zone> zone(new msgpack::zone);
-#pragma GCC diagnostic pop
-    msgpack::unpack_return rv = msgpack::unpack(data, size, &offset, zone.get(), &object);
+    std::vector<char> buffer(size);
+    memcpy(buffer.data(), data, size);
+    msgpack::unpack_return rv = msgpack::unpack(buffer.data(), size, &offset, &zone, &object);
```
Comment (3Hren) [inline src/decoder.cpp]: Check for `zone` lifetime and functionality.

**Comment by 3Hren** [comment]
LGTM, except some issues.

Fix them, and after that I'll pass Unit Tests through Valgrind and merge (if ok).

File: src/basic_session.cpp (hunk 7 above)
Comment (antmat) [inline src/basic_session.cpp]: to distinguish with encoder_t. I'll find a better name.

File: include/cocaine/framework/sender.hpp (hunk 4 above)
Comment (antmat) [inline include/cocaine/framework/sender.hpp]: ok, but what's wrong with it? It's located in function scope, so does not affect outer namespace.

**Pull request merged**